server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1579,6 +1597,64 @@ ChannelFactory getChannelFactory() {
         }
     }
 
+    public void testTranslogOperationListener() throws IOException {
+        Path tempDir = createTempDir();
+        final Settings settings = Settings.builder().put(IndexMetadata.SETTING_VERSION_CREATED, org.elasticsearch.Version.CURRENT).build();
+
+        final ArrayList<Long> seqNos = new ArrayList<>();
+        final ArrayList<Location> locations = new ArrayList<>();
+        final ArrayList<BytesReference> datas = new ArrayList<>();
+        OperationListener listener = (data, seqNo, location) -> {
+            seqNos.add(seqNo);
+            locations.add(location);
+            try (BytesStreamOutput output = new BytesStreamOutput()) {
+                try {
+                    data.writeTo(output);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+                datas.add(output.bytes());
+            }
+
+        };
+        TranslogConfig config = getTranslogConfig(tempDir, settings, listener);
+        Translog translog = createTranslog(config);
+
+        Location location1 = translog.add(indexOp(randomAlphaOfLength(10), 0, primaryTerm.get()));
+        Location location2 = translog.add(TranslogOperationsUtils.indexOp(randomAlphaOfLength(10), 1, primaryTerm.get()));
+        long firstGeneration = translog.getGeneration().translogFileGeneration;
+        assertThat(location1.generation, equalTo(firstGeneration));
+        assertThat(location2.generation, equalTo(firstGeneration));
+
+        translog.rollGeneration();
+
+        Location location3 = translog.add(TranslogOperationsUtils.indexOp(randomAlphaOfLength(10), 3, primaryTerm.get()));
+        Location location4 = translog.add(TranslogOperationsUtils.indexOp(randomAlphaOfLength(10), 2, primaryTerm.get()));
+        long secondGeneration = translog.getGeneration().translogFileGeneration;
+        assertThat(location3.generation, equalTo(secondGeneration));
+        assertThat(location4.generation, equalTo(secondGeneration));
+
+        try {
+            assertThat(seqNos.get(0), equalTo(0L));
```

Comment:
I think we could compare the collections as a whole without doing element by element
```java
assertThat(seqNos, equalTo(List.of(0L, 1L, 3L, 2L)));
assertThat(locations, equalTo(List.of(location1, location2, location3, location4)));
```